### ISSUE TITLE
OneIntegerGenericBitSet.clear is not implemented

### ISSUE THREAD

**Issue by kheradmand** [kframework/k-legacy#2273]
I got the following assertion : 
```
java.lang.AssertionError: unimplemented
	at org.kframework.utils.OneIntegerGenericBitSet.clear(OneIntegerGenericBitSet.java:183)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.addUnification(FastRuleMatcher.java:515)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.match(FastRuleMatcher.java:228)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.matchAssoc(FastRuleMatcher.java:400)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.matchAssoc(FastRuleMatcher.java:441)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.match(FastRuleMatcher.java:323)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.match(FastRuleMatcher.java:271)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.match(FastRuleMatcher.java:315)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.match(FastRuleMatcher.java:315)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.matchRulePattern(FastRuleMatcher.java:107)
	at org.kframework.backend.java.symbolic.SymbolicRewriter.fastComputeRewriteStep(SymbolicRewriter.java:164)
	at org.kframework.backend.java.symbolic.SymbolicRewriter.computeRewriteStep(SymbolicRewriter.java:105)
	at org.kframework.backend.java.symbolic.SymbolicRewriter.rewrite(SymbolicRewriter.java:86)
	at org.kframework.backend.java.symbolic.InitializeRewriter$SymbolicRewriterGlue.execute(InitializeRewriter.java:132)
	at org.kframework.krun.modes.KRunExecutionMode.execute(KRunExecutionMode.java:65)
	at org.kframework.krun.KRun.run(KRun.java:95)
	at org.kframework.krun.KRunFrontEnd.run(KRunFrontEnd.java:68)
	at org.kframework.main.FrontEnd.main(FrontEnd.java:34)
	at org.kframework.main.Main.runApplication(Main.java:420)
	at org.kframework.main.Main.runApplication(Main.java:271)
	at org.kframework.main.Main.main(Main.java:84)
```

And  OneIntegerGenericBitSet.clear is really not implemented:
```
    public void clear(int i) {
        throw new AssertionError("unimplemented");
    }
```

**Comment by kheradmand** [comment]
@daejunpark @andreistefanescu 
Actually what is  `OneIntegerGenericBitSet`  and why it exists ? Why not java.util.BitSet?

**Comment by kheradmand** [comment]
There is this comment at the top of the class declaration:
```
/**
 * {@link BitSet} implementation special-casing for just one integer in the set. Not particularly good so could
 * delete at some point.
 */
```

So maybe it would be better to get rid of it.

**Comment by daejunpark** [comment]
@kheradmand Why don't you propose a solution and make a pull request for it?

**Comment by kheradmand** [comment]
@daejunpark  I replace the `OneIntegerGenericBitSet` with a wrapper for `java.util.BitSet`. Please take a look at 8141c5b2e2f97a8ef23dcb3d28b33e7ed051c4b6. Now I get this exception:
```
java.lang.StackOverflowError
	at org.kframework.backend.java.kil.BuiltinList.size(BuiltinList.java:181)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.matchAssoc(FastRuleMatcher.java:367)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.match(FastRuleMatcher.java:323)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.matchAssoc(FastRuleMatcher.java:438)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.match(FastRuleMatcher.java:323)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.matchAssoc(FastRuleMatcher.java:438)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.match(FastRuleMatcher.java:323)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.matchAssoc(FastRuleMatcher.java:438)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.match(FastRuleMatcher.java:323)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.matchAssoc(FastRuleMatcher.java:438)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.match(FastRuleMatcher.java:323)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.matchAssoc(FastRuleMatcher.java:438)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.match(FastRuleMatcher.java:323)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.matchAssoc(FastRuleMatcher.java:438)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.match(FastRuleMatcher.java:323)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.matchAssoc(FastRuleMatcher.java:438)
```

I don't know if it is a bug in other parts of K, or is related to my change.

**Comment by kheradmand** [comment]
That problem was due to a bug in my wrapper class, I fixed that (see 57fff53ffc1eb0f7fcca6d542ad5754336f5bacb).
But now I get another kind of error:
```
(error "line 5 column 27: invalid function application for =, sort mismatch on argument at position 2, expected Bool but given BoolExpr")
(error "line 5 column 43: invalid function application for =, sort mismatch on argument at position 2, expected Bool but given BoolExpr")
[Warning] Critical: failed to translate smtlib expression:
(declare-sort BoolExpr)
(declare-fun _2825 () Bool)
(declare-fun _2823 () Bool)
(declare-fun _2824 () BoolExpr)
(assert (and (= _2823 _2824) (= _2825 false)))
[Warning] Critical: failed to translate smtlib expression:
(declare-sort BoolExpr)
(declare-fun _2826 () Bool)
(declare-fun _2827 () Bool)
(declare-fun _2828 () BoolExpr)
(assert (and (= _2826 false) (= _2827 _2828)))
```

**Comment by daejunpark** [comment]
This type of errors happens mostly when it attempts to apply a wrong rule. You can try to find the rule to be applied at that point, and locate (sub-)terms of sort `BoolExpr`.

**Comment by kheradmand** [comment]
@daejunpark 
I had these two rules: 
```
rule <k> @resolveArgs((X:Id , Rest:Args) => Rest,$resolved(_:List (.List => ListItem(V)))) ... </k>
     <frameStack> ListItem((X |-> V:Arg) _:Map) _:List </frameStack>
rule <k> @resolveArgs((X:Id , Rest:Args) => Rest,$resolved(_:List (.List => ListItem(X)))) ... </k>
     <frameStack> ListItem(M:Map) _:List </frameStack> when notBool (X in_keys(M))
```

But it seemed that the second rule gets applied even if `notBool (X in_keys(M))` is not true (i.e even when `X` have an associated entry at the top of `frameStack`, `X` itself gets added to `$resolve` instead of `V`)


So I wanted to check what was the value of `notBool (X in_keys(M))`,  so I changed to second rule to:
```
rule <k> @resolveArgs((X:Id , Rest:Args),$resolved(_:List )) => (notBool (X in_keys(M))) ... </k>
     <frameStack> ListItem(M:Map) _:List </frameStack> //when notBool (X in_keys(M))
```

But it caused some ambiguities,
```
[Warning] Inner Parser: Parsing ambiguity. Arbitrarily choosing the first.
1: syntax Bool@BOOL-SYNTAX ::= "(" Bool@BOOL-SYNTAX ")"
[Location(#token("214",Int@INT-SYNTAX),#token("19",Int@INT-SYNTAX),#token("214",Int@INT-SYNTAX),#token("40",Int@INT-SYNTAX))
Source(#token("/home/ali/FSL/k/k-distribution/target/release/k/lib/java/../../include/builtin/domains.k",KString@KSTRING))
bracket(#token("",AttributeValue)) klabel(#token("(_)",KString@KSTRING))]
   
(_)(notBool_(amb((_)(_in_keys(_)(#SemanticCastToId@ID(#token(#KVariable@KSEQ-SYMBOLIC,"X")),#SemanticCastToMap@MAP(#token(#KVariable@KSEQ-SYMBOLIC,"M")))),(_)(_in_keys(_)(#SemanticCastToId@ID(#token(#KVariable@KSEQ-SYMBOLIC,"X")),#SemanticCastToMap@MAP(#token(#KVariable@KSEQ-SYMBOLIC,"M")))))))
2: syntax BoolExpr@CONTROL-SYNTAX ::= "(" BoolExpr@CONTROL-SYNTAX ")"
[Location(#token("63",Int@INT-SYNTAX),#token("23",Int@INT-SYNTAX),#token("63",Int@INT-SYNTAX),#token("64",Int@INT-SYNTAX))
Source(#token("/home/ali/P4/p4k/src/./syntax/control-syntax.k",KString@KSTRING))
braket(#token("",AttributeValue)) klabel(#token("(_)",KString@KSTRING))]
   
(_)(notBool_(amb((_)(_in_keys(_)(#SemanticCastToId@ID(#token(#KVariable@KSEQ-SYMBOLIC,"X")),#SemanticCastToMap@MAP(#token(#KVariable@KSEQ-SYMBOLIC,"M")))),(_)(_in_keys(_)(#SemanticCastToId@ID(#token(#KVariable@KSEQ-SYMBOLIC,"X")),#SemanticCastToMap@MAP(#token(#KVariable@KSEQ-SYMBOLIC,"M")))))))
3: syntax KBott@BASIC-K ::= "(" K@SORT-K ")"
[Location(#token("63",Int@INT-SYNTAX),#token("24",Int@INT-SYNTAX),#token("63",Int@INT-SYNTAX),#token("45",Int@INT-SYNTAX))
Source(#token("/home/ali/FSL/k/k-distribution/target/release/k/lib/java/../../include/builtin/kast.k",KString@KSTRING))
bracket(#token("",AttributeValue)) klabel(#token("(_)",KString@KSTRING))]
   
(_)(notBool_(amb((_)(_in_keys(_)(#SemanticCastToId@ID(#token(#KVariable@KSEQ-SYMBOLIC,"X")),#SemanticCastToMap@MAP(#token(#KVariable@KSEQ-SYMBOLIC,"M")))),(_)(_in_keys(_)(#SemanticCastToId@ID(#token(#KVariable@KSEQ-SYMBOLIC,"X")),#SemanticCastToMap@MAP(#token(#KVariable@KSEQ-SYMBOLIC,"M")))))))
	Source(./configuration.k)
	Location(613,65,613,89)
```

 so I changed it to:
```
rule <k> @resolveArgs((X:Id , Rest:Args),$resolved(_:List )) => $temp(notBool (X in_keys(M))) ... </k>
     <frameStack> ListItem(M:Map) _:List </frameStack> //when notBool (X in_keys(M))


syntax KItem ::= "$temp" "(" Bool ")"
```

But it didn't get evaluated:
```
<T> <k> $temp ( notBool ( meter_idx in_keys ( meter_idx |-> @val ( 10 , 0 , false ) ) ) ) ~> #freezer@call(_,_)0  ...
```

so I changed the rule to:
```
rule <k> @resolveArgs((X:Id , Rest:Args),$resolved(_:List )) => $temp(notBool (X in_keys(M))) ... </k>
     <frameStack> ListItem(M:Map) _:List </frameStack> //when notBool (X in_keys(M))


syntax KItem ::= "$temp" "(" K ")" [strict]
```

again the same:
```
<T> <k> $temp ( notBool ( meter_idx in_keys ( meter_idx |-> @val ( 10 , 0 , false ) ) ) ) ~> #freezer@call(_,_)0  ...
```

so I changed it to:
```
rule <k> @resolveArgs((X:Id , Rest:Args),$resolved(_:List )) => $temp(notBool (X in_keys(M))) ... </k>
     <frameStack> ListItem(M:Map) _:List </frameStack> //when notBool (X in_keys(M))


syntax KItem ::= "$temp" "(" K ")" [strict]
rule $temp(B:Bool) => B
```

which results in that error:
```
(error "line 5 column 27: invalid function application for =, sort mismatch on argument at position 2, expected Bool but given BoolExpr")
(error "line 5 column 27: invalid function application for =, sort mismatch on argument at position 2, expected Bool but given BoolExpr")
(error "line 5 column 43: invalid function application for =, sort mismatch on argument at position 2, expected Bool but given BoolExpr")
java.lang.AssertionError
	at org.kframework.backend.java.symbolic.ConjunctiveFormula.getSubstitution(ConjunctiveFormula.java:534)
	at org.kframework.backend.java.symbolic.ConjunctiveFormula.simplify(ConjunctiveFormula.java:417)
	at org.kframework.backend.java.symbolic.ConjunctiveFormula.simplify(ConjunctiveFormula.java:323)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.addUnification(FastRuleMatcher.java:512)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.match(FastRuleMatcher.java:228)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.matchAssoc(FastRuleMatcher.java:438)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.match(FastRuleMatcher.java:323)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.match(FastRuleMatcher.java:271)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.match(FastRuleMatcher.java:315)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.match(FastRuleMatcher.java:315)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.matchRulePattern(FastRuleMatcher.java:107)
	at org.kframework.backend.java.symbolic.SymbolicRewriter.fastComputeRewriteStep(SymbolicRewriter.java:164)
	at org.kframework.backend.java.symbolic.SymbolicRewriter.computeRewriteStep(SymbolicRewriter.java:105)
	at org.kframework.backend.java.symbolic.SymbolicRewriter.rewrite(SymbolicRewriter.java:86)
	at org.kframework.backend.java.symbolic.InitializeRewriter$SymbolicRewriterGlue.execute(InitializeRewriter.java:148)
	at org.kframework.krun.modes.KRunExecutionMode.execute(KRunExecutionMode.java:65)
	at org.kframework.krun.KRun.run(KRun.java:103)
	at org.kframework.krun.KRunFrontEnd.run(KRunFrontEnd.java:68)
	at org.kframework.main.FrontEnd.main(FrontEnd.java:34)
	at org.kframework.main.Main.runApplication(Main.java:414)
	at org.kframework.main.Main.runApplication(Main.java:265)
	at org.kframework.main.Main.main(Main.java:74)
java.lang.AssertionError
	at org.kframework.backend.java.symbolic.ConjunctiveFormula.getSubstitution(ConjunctiveFormula.java:534)
	at org.kframework.backend.java.symbolic.ConjunctiveFormula.simplify(ConjunctiveFormula.java:417)
	at org.kframework.backend.java.symbolic.ConjunctiveFormula.simplify(ConjunctiveFormula.java:323)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.addUnification(FastRuleMatcher.java:512)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.match(FastRuleMatcher.java:228)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.matchAssoc(FastRuleMatcher.java:438)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.match(FastRuleMatcher.java:323)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.match(FastRuleMatcher.java:271)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.match(FastRuleMatcher.java:315)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.match(FastRuleMatcher.java:315)
	at org.kframework.backend.java.symbolic.FastRuleMatcher.matchRulePattern(FastRuleMatcher.java:107)
	at org.kframework.backend.java.symbolic.SymbolicRewriter.fastComputeRewriteStep(SymbolicRewriter.java:164)
	at org.kframework.backend.java.symbolic.SymbolicRewriter.computeRewriteStep(SymbolicRewriter.java:105)
	at org.kframework.backend.java.symbolic.SymbolicRewriter.rewrite(SymbolicRewriter.java:86)
	at org.kframework.backend.java.symbolic.InitializeRewriter$SymbolicRewriterGlue.execute(InitializeRewriter.java:148)
	at org.kframework.krun.modes.KRunExecutionMode.execute(KRunExecutionMode.java:65)
	at org.kframework.krun.KRun.run(KRun.java:103)
	at org.kframework.krun.KRunFrontEnd.run(KRunFrontEnd.java:68)
	at org.kframework.main.FrontEnd.main(FrontEnd.java:34)
	at org.kframework.main.Main.runApplication(Main.java:414)
	at org.kframework.main.Main.runApplication(Main.java:265)
	at org.kframework.main.Main.main(Main.java:74)
[Error] Internal: Uncaught exception thrown of type AssertionError.
Please rerun your program with the --debug flag to generate a stack trace, and
file a bug report at https://github.com/kframework/k/issues
[Warning] Critical: failed to translate smtlib expression:
(declare-sort BoolExpr)
(declare-fun _2807 () Bool)
(declare-fun _2809 () Bool)
(declare-fun _2808 () BoolExpr)
(assert (and (= _2807 _2808) (= _2809 false)))
[Warning] Critical: failed to translate smtlib expression:
(declare-sort BoolExpr)
(declare-fun _2810 () Bool)
(declare-fun _2812 () Bool)
(declare-fun _2811 () BoolExpr)
(assert (and (= _2810 _2811) (= _2812 false)))
[Warning] Critical: failed to translate smtlib expression:
(declare-sort BoolExpr)
(declare-fun _2827 () Bool)
(declare-fun _2828 () BoolExpr)
(declare-fun _2826 () Bool)
(assert (and (= _2826 false) (= _2827 _2828)))

```

**Comment by daejunpark** [comment]
try:
```
X in keys(M)
```
note that there is no underscore (`_`).

**Comment by kheradmand** [comment]
@daejunpark 
It works! But why that one (in_keys) wasn't working?

**Comment by kheradmand** [comment]
Oh I just saw your pull request

**Comment by kheradmand** [comment]
fixed in #2274 (and #2275)
